zerver/tests/test_service_bot_system.py

```diff
@@ -177,6 +179,88 @@ def test_entry_removal(self) -> None:
+        self.assert_json_success(result)
+        self.assertEqual(result.json()['state'], {'key 3': 'value 3', 'key 1': 'value 1'})
+        # Assert the stored data for all keys.
+        result = self.client_get('/json/user_state')
+        self.assert_json_success(result)
+        self.assertEqual(result.json()['state'], initial_dict)
+        # Store some more data; update an entry and store a new entry
+        dict_update = {'key 1': 'new value', 'key 4': 'value 4'}
+        params = {
+            'state': ujson.dumps(dict_update)
+        }
+        result = self.client_put('/json/user_state', params)
+        self.assert_json_success(result)
+        # Assert the data was updated.
+        updated_dict = initial_dict.copy()
+        updated_dict.update(dict_update)
+        result = self.client_get('/json/user_state')
+        self.assert_json_success(result)
+        self.assertEqual(result.json()['state'], updated_dict)
+        # Assert errors on invalid requests.
+        params = {  # type: ignore # Ignore 'incompatible type "str": "List[str]"; expected "str": "str"' for testing
+            'keys': ["This is a list, but should be a serialized string."]
+        }
+        result = self.client_get('/json/user_state', params)
+        self.assert_json_error(result, 'Argument "keys" is not valid JSON.')
+        params = {
+            'keys': ujson.dumps(["key 1", "nonexistent key"])
+        }
+        result = self.client_get('/json/user_state', params)
+        self.assert_json_error(result, "Cannot get state. <UserProfile: othello@zulip.com <Realm: zulip 1>> "
+                                       "doesn't have an entry with the key 'nonexistent key'.")
+        params = {  # type: ignore # Ignore 'incompatible type "str": "List[str]"; expected "str": "str"' for testing
+            'state': ujson.dumps({'foo': [1, 2, 3]})
+        }
+        result = self.client_put('/json/user_state', params)
+        self.assert_json_error(result, "Cannot set state. The value type is <class 'list'>, but it should be str.")
+        # Remove some entries.
+        keys_to_remove = ['key 1', 'key 2']
+        params = {
+            'keys': ujson.dumps(keys_to_remove)
+        }
+        result = self.client_delete('/json/user_state', params)
+        self.assert_json_success(result)
+        # Assert the entries were removed.
+        for key in keys_to_remove:
+            updated_dict.pop(key)
+        result = self.client_get('/json/user_state')
+        self.assert_json_success(result)
+        self.assertEqual(result.json()['state'], updated_dict)
+        # Try to remove an existing and a nonexistent key.
+        params = {
+            'keys': ujson.dumps(['key 3', 'nonexistent key'])
+        }
+        result = self.client_delete('/json/user_state', params)
+        self.assert_json_error(result, "Cannot remove state. Key 'nonexistent key' does not exist. No keys were removed.")
```

Comment:
I think just "Key does not exist." or if you want to be more verbose "Key '%s' does not exist" is clear enough.  It'll be clear what they were trying to do from the surrounding code in whatever is calling this endpoint, I think.

Reply:
(And I think it's probably fine to cover that no keys were removed in the docs; that's what one should expect a well-defined API to do...)